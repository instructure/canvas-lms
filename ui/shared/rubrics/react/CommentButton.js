/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import PropTypes from 'prop-types'
import {Button} from '@instructure/ui-buttons'
import {IconFeedbackLine} from '@instructure/ui-icons'
import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import I18n from 'i18n!edit_rubricCommentButton'

const CommentButton = ({onClick}) => (
  <div>
    <Button variant="icon" icon={<IconFeedbackLine />} margin="0 x-small 0 0" onClick={onClick}>
      <ScreenReaderContent>{I18n.t('Add Additional Comments')}</ScreenReaderContent>
    </Button>
  </div>
)
CommentButton.propTypes = {
  onClick: PropTypes.func.isRequired
}

export default CommentButton
