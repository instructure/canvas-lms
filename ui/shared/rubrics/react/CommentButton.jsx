/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import PropTypes from 'prop-types'
import {IconButton} from '@instructure/ui-buttons'
import {IconFeedbackLine} from '@instructure/ui-icons'
import {useScope as createI18nScope} from '@canvas/i18n'

const I18n = createI18nScope('edit_rubricCommentButton')

const CommentButton = ({onClick}) => (
  <div className="add-comment-button-wrapper">
    <IconButton
      renderIcon={<IconFeedbackLine />}
      withBackground={false}
      withBorder={false}
      margin="0 x-small 0 0"
      onClick={onClick}
      screenReaderLabel={I18n.t('Add Additional Comments')}
      data-testid="add-comment-button"
    />
  </div>
)
CommentButton.propTypes = {
  onClick: PropTypes.func.isRequired,
}

export default CommentButton
