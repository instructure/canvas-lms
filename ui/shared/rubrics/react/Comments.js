/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import PropTypes from 'prop-types'

import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import {Text} from '@instructure/ui-text'
import {Checkbox} from '@instructure/ui-checkbox'
import {SimpleSelect} from '@instructure/ui-simple-select'
import {TextArea} from '@instructure/ui-text-area'
import I18n from 'i18n!edit_rubricComments'

import {assessmentShape} from './types'

const ellipsis = () => I18n.t('â€¦')

const truncate = comment => (comment.length > 100 ? comment.slice(0, 99) + ellipsis() : comment)

const slug = str => str.replace(/\W/g, '')

const FreeFormComments = props => {
  const {allowSaving, savedComments, comments, large, saveLater, setComments, setSaveLater} = props
  const first = (
    <SimpleSelect.Option key="first" id="first" value="first">
      {I18n.t('[ Select ]')}
    </SimpleSelect.Option>
  )

  const options = savedComments.map((comment, ix) => (
    <SimpleSelect.Option
      key={slug(comment).slice(-8)}
      id={`${slug(comment).slice(-6)}_${ix}`}
      value={ix.toString()}
      label={comment}
    >
      {truncate(comment)}
    </SimpleSelect.Option>
  ))
  const selector = [
    <SimpleSelect
      renderLabel={I18n.t('Saved Comments')}
      assistiveText={I18n.t('Select from saved comments')}
      onChange={(_unused, el) => {
        setComments(savedComments[el.value])
      }}
    >
      {[first, ...options]}
    </SimpleSelect>,
    <br key="br" />
  ]

  const saveBox = () => {
    if (allowSaving && large) {
      return (
        <Checkbox
          checked={saveLater}
          label={I18n.t('Save this comment for reuse')}
          size="small"
          onChange={event => setSaveLater(event.target.checked)}
        />
      )
    }
  }

  const label = I18n.t('Comments')
  const toScreenReader = el => <ScreenReaderContent>{el}</ScreenReaderContent>

  const commentClass = `edit-freeform-comments-${large ? 'large' : 'small'}`
  return (
    <div className={commentClass}>
      {options.length > 0 ? selector : null}
      <TextArea
        data-selenium="criterion_comments_text"
        label={large ? label : toScreenReader(label)}
        placeholder={large ? undefined : label}
        maxHeight="50rem"
        onChange={e => setComments(e.target.value)}
        resize="vertical"
        size="small"
        value={comments}
      />
      {large ? <br /> : null}
      {saveBox()}
    </div>
  )
}
FreeFormComments.propTypes = {
  allowSaving: PropTypes.bool.isRequired,
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  comments: PropTypes.string,
  large: PropTypes.bool.isRequired,
  saveLater: PropTypes.bool,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
FreeFormComments.defaultProps = {
  comments: '',
  saveLater: false
}

const commentElement = assessment => {
  if (assessment.comments_html || assessment.comments) {
    return (
      <div>
        <Text size="small" weight="bold">
          {I18n.t('Comments')}
        </Text>
        {assessment.comments_html ? (
          <div dangerouslySetInnerHTML={{__html: assessment.comments_html}} />
        ) : (
          <div>{assessment.comments}</div>
        )}
      </div>
    )
  } else {
    return null
  }
}

export const CommentText = ({assessment, placeholder, weight}) => (
  <span className="react-rubric-break-words">
    <Text size="x-small" weight={weight}>
      {assessment !== null ? commentElement(assessment) : placeholder}
    </Text>
  </span>
)
CommentText.propTypes = {
  assessment: PropTypes.shape(assessmentShape),
  placeholder: PropTypes.string,
  weight: PropTypes.string.isRequired
}
CommentText.defaultProps = {
  assessment: null,
  placeholder: ''
}

const Comments = props => {
  const {editing, assessment, footer, ...commentProps} = props
  if (!editing || assessment === null) {
    return (
      <div className="rubric-freeform">
        <CommentText
          assessment={assessment}
          placeholder={I18n.t(
            'This area will be used by the assessor to leave comments related to this criterion.'
          )}
          weight="normal"
        />
        {footer}
      </div>
    )
  } else {
    return (
      <FreeFormComments
        comments={assessment.comments}
        saveLater={assessment.saveCommentsForLater}
        {...commentProps}
      />
    )
  }
}
Comments.propTypes = {
  allowSaving: PropTypes.bool,
  editing: PropTypes.bool.isRequired,
  assessment: PropTypes.shape(assessmentShape),
  footer: PropTypes.node,
  large: PropTypes.bool,
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
Comments.defaultProps = {
  allowSaving: true,
  footer: null,
  large: true
}

export default Comments
