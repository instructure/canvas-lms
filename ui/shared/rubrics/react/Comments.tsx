/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import type {ReactNode} from 'react'

import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import {Text} from '@instructure/ui-text'
import {Checkbox} from '@instructure/ui-checkbox'
import {SimpleSelect} from '@instructure/ui-simple-select'
import {TextArea} from '@instructure/ui-text-area'
import {useScope as createI18nScope} from '@canvas/i18n'

const I18n = createI18nScope('edit_rubricComments')

const ellipsis = () => I18n.t('â€¦')

const truncate = (comment: string) =>
  comment.length > 100 ? comment.slice(0, 99) + ellipsis() : comment

const slug = (str: string) => str.replace(/\W/g, '')

interface Assessment {
  criterion_id: string
  comments?: string
  comments_html?: string
  points: {
    text?: string
    value?: number
    valid?: boolean
  }
  focusPoints?: number
  saveCommentsForLater?: boolean
}

interface FreeFormCommentsProps {
  allowSaving: boolean
  savedComments: string[]
  comments?: string
  large: boolean
  saveLater?: boolean
  setComments: (comments: string) => void
  setSaveLater: (saveLater: boolean) => void
}

const FreeFormComments = ({
  allowSaving,
  savedComments,
  comments = '',
  large,
  saveLater = false,
  setComments,
  setSaveLater,
}: FreeFormCommentsProps) => {
  const first = (
    <SimpleSelect.Option key="first" id="first" value="first">
      {I18n.t('[ Select ]')}
    </SimpleSelect.Option>
  )

  const options = savedComments.map((comment, ix) => (
    <SimpleSelect.Option
      key={slug(comment).slice(-8)}
      id={`${slug(comment).slice(-6)}_${ix}`}
      value={ix.toString()}
      label={comment}
    >
      {truncate(comment)}
    </SimpleSelect.Option>
  ))
  const selector = [
    <SimpleSelect
      renderLabel={I18n.t('Saved Comments')}
      assistiveText={I18n.t('Select from saved comments')}
      onChange={(_unused: any, el: any) => {
        setComments(savedComments[el.value])
      }}
      key="simple-select"
    >
      {[first, ...options]}
    </SimpleSelect>,
    <br key="br" />,
  ]

  const saveBox = () => {
    if (allowSaving && large) {
      return (
        <Checkbox
          checked={saveLater}
          label={I18n.t('Save this comment for reuse')}
          size="small"
          onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
            setSaveLater(event.target.checked)
          }
        />
      )
    }
  }

  const label = I18n.t('Comments')
  const toScreenReader = (el: string) => <ScreenReaderContent>{el}</ScreenReaderContent>

  const commentClass = `rubric-comment edit-freeform-comments-${large ? 'large' : 'small'}`
  return (
    <div className={commentClass}>
      {options.length > 0 ? selector : null}
      <TextArea
        data-selenium="criterion_comments_text"
        label={large ? label : toScreenReader(label)}
        placeholder={large ? undefined : label}
        maxHeight="50rem"
        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => setComments(e.target.value)}
        resize="vertical"
        size="small"
        value={comments}
      />
      {large ? <br /> : null}
      {saveBox()}
    </div>
  )
}

const commentElement = (assessment: Assessment) => {
  if (assessment.comments_html || assessment.comments) {
    return (
      <div>
        <Text size="small" weight="bold">
          {I18n.t('Comments')}
        </Text>
        {assessment.comments_html ? (
          <div dangerouslySetInnerHTML={{__html: assessment.comments_html}} />
        ) : (
          <div>{assessment.comments}</div>
        )}
      </div>
    )
  } else {
    return null
  }
}

interface CommentTextProps {
  assessment?: Assessment | null
  placeholder?: string
  weight: 'normal' | 'light' | 'bold'
}

export const CommentText = ({assessment = null, placeholder = '', weight}: CommentTextProps) => (
  <span className="react-rubric-break-words">
    <Text size="x-small" weight={weight}>
      {assessment !== null ? commentElement(assessment) : placeholder}
    </Text>
  </span>
)

interface CommentsProps {
  allowSaving?: boolean
  editing: boolean
  assessment?: Assessment | null
  footer?: ReactNode
  large?: boolean
  savedComments: string[]
  setComments: (comments: string) => void
  setSaveLater: (saveLater: boolean) => void
}

const Comments = ({
  allowSaving = true,
  editing,
  assessment,
  footer = null,
  large = true,
  savedComments,
  setComments,
  setSaveLater,
}: CommentsProps) => {
  if (!editing || assessment === null) {
    return (
      <div className="rubric-freeform">
        <CommentText
          assessment={assessment}
          placeholder={I18n.t(
            'This area will be used by the assessor to leave comments related to this criterion.',
          )}
          weight="normal"
        />
        {footer}
      </div>
    )
  } else {
    return (
      <FreeFormComments
        comments={assessment?.comments}
        saveLater={assessment?.saveCommentsForLater}
        allowSaving={allowSaving}
        savedComments={savedComments}
        large={large}
        setComments={setComments}
        setSaveLater={setSaveLater}
      />
    )
  }
}

export default Comments
