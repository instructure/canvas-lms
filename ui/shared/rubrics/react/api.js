/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/*
  TODO: Duplicated and modified within jsx/outcomes/MasteryScale for use there
        Remove when feature flag account_level_mastery_scales is enabled
*/

import axios from '@canvas/axios'

export const fetchProficiency = accountId =>
  axios.get(`/api/v1/accounts/${accountId}/outcome_proficiency`)

export const saveProficiency = (accountId, config) =>
  axios.post(`/api/v1/accounts/${accountId}/outcome_proficiency`, config)
