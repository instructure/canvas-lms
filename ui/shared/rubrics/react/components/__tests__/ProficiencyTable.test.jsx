/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/*
  TODO: Duplicated and modified within jsx/outcomes/MasteryScale for use there
        Remove when feature flag account_level_mastery_scales is enabled
*/

import _ from 'lodash'
import $ from 'jquery'
import React from 'react'
import axios from '@canvas/axios'
import {render, waitFor, fireEvent} from '@testing-library/react'
import ProficiencyTable from '../ProficiencyTable'

// Suppress the validateDOMNesting warning for this test suite
// The ProficiencyTable uses Flex.Item with as="th" which causes this warning
const originalError = console.error
beforeAll(() => {
  console.error = (...args) => {
    if (typeof args[0] === 'string' && args[0].includes('validateDOMNesting')) {
      return
    }
    originalError.call(console, ...args)
  }
})

afterAll(() => {
  console.error = originalError
})

// Mock HTMLElement.focus to prevent focus errors in tests
beforeEach(() => {
  HTMLElement.prototype.focus = jest.fn()
  jest.useFakeTimers()
})

afterEach(() => {
  jest.useRealTimers()
})

async function wait(ms = 0) {
  if (ms > 0) {
    jest.advanceTimersByTime(ms)
  }
  // Allow promises to resolve
  await Promise.resolve()
}

const defaultProps = {
  accountId: '1',
}

let getSpy

describe('default proficiency', () => {
  beforeEach(() => {
    const err = _.assign(new Error(), {response: {status: 404}})
    getSpy = jest.spyOn(axios, 'get').mockImplementation(() => Promise.reject(err))
  })

  afterEach(() => {
    getSpy.mockRestore()
  })

  it('renders loading spinner initially', () => {
    const {getByText} = render(<ProficiencyTable {...defaultProps} />)
    expect(getByText('Loading')).toBeInTheDocument()
  })

  it('render billboard after loading', async () => {
    const {getByText} = render(<ProficiencyTable {...defaultProps} />)
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
  })
  it('renders five ratings', async () => {
    const {getAllByRole, getByText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    await waitFor(() => {
      const checkboxes = getAllByRole('radio')
      expect(checkboxes).toHaveLength(5) // Each rating has a mastery radio button
    })
  })

  it('has mastery selected on first rating only', async () => {
    const {getAllByRole, getByText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    await waitFor(() => {
      const masteryRadios = getAllByRole('radio')
      expect(masteryRadios).toHaveLength(5)
      // Default mastery index is 1 (second rating)
      expect(masteryRadios[0]).not.toBeChecked()
      expect(masteryRadios[1]).toBeChecked()
      expect(masteryRadios[2]).not.toBeChecked()
      expect(masteryRadios[3]).not.toBeChecked()
      expect(masteryRadios[4]).not.toBeChecked()
    })
  })

  it('clicking add button adds rating', async () => {
    const {getAllByRole, getByText, getByRole} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    await waitFor(() => {
      const initialRadios = getAllByRole('radio')
      expect(initialRadios).toHaveLength(5)
    })

    const addButton = getByRole('button', {name: /add proficiency rating/i})
    fireEvent.click(addButton)

    await waitFor(() => {
      const newRadios = getAllByRole('radio')
      expect(newRadios).toHaveLength(6)
    })
  })

  it('clicking add rating button flashes SR message', async () => {
    const {getByText, getByRole} = render(<ProficiencyTable {...defaultProps} />)
    const flashMock = jest.spyOn($, 'screenReaderFlashMessage')

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Wait for table to render
    await waitFor(() => {
      expect(getByText('Proficiency Rating')).toBeInTheDocument()
    })

    const addButton = getByRole('button', {name: 'Add proficiency rating'})
    fireEvent.click(addButton)

    expect(flashMock).toHaveBeenCalledTimes(1)
    flashMock.mockRestore()
  })

  it('deleting rating removes rating and flashes SR message', async () => {
    const {getAllByRole, getByText} = render(<ProficiencyTable {...defaultProps} />)
    const flashMock = jest.spyOn($, 'screenReaderFlashMessage')

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    await waitFor(() => {
      const initialRadios = getAllByRole('radio')
      expect(initialRadios).toHaveLength(5)
    })

    const deleteButtons = getAllByRole('button', {name: /delete proficiency rating/i})
    fireEvent.click(deleteButtons[1])

    await waitFor(() => {
      const newRadios = getAllByRole('radio')
      expect(newRadios).toHaveLength(4)
    })
    expect(flashMock).toHaveBeenCalledTimes(1)
    flashMock.mockRestore()
  })

  it('setting blank description sets error', async () => {
    const {getByText, getAllByRole} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Wait for inputs to be rendered
    await waitFor(() => {
      expect(getByText('Proficiency Rating')).toBeInTheDocument()
    })

    // Clear first rating description
    const descriptionInputs = getAllByRole('textbox', {name: /change description/i})
    const firstInput = descriptionInputs[0]
    fireEvent.change(firstInput, {target: {value: ''}})
    fireEvent.blur(firstInput)

    // Submit form
    const saveButton = getByText('Save Learning Mastery')
    fireEvent.click(saveButton)

    await waitFor(() => {
      expect(getByText(/Please include a rating title/)).toBeInTheDocument()
    })
  })

  it('setting blank points sets error', async () => {
    const {getByText, getAllByLabelText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Clear first rating points
    const pointsInputs = getAllByLabelText(/Change points/)
    const firstInput = pointsInputs[0]
    firstInput.focus()
    firstInput.select()
    fireEvent.input(firstInput, {target: {value: ''}})

    // Submit form
    const saveButton = getByText('Save Learning Mastery')
    fireEvent.click(saveButton)

    await waitFor(() => {
      expect(getByText(/Invalid format/)).toBeInTheDocument()
    })
  })

  it('setting invalid points sets error', async () => {
    const {getByText, getAllByLabelText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Set invalid points
    const pointsInputs = getAllByLabelText(/Change points/)
    const firstInput = pointsInputs[0]
    firstInput.focus()
    firstInput.select()
    fireEvent.input(firstInput, {target: {value: '1.1.1'}})

    // Submit form
    const saveButton = getByText('Save Learning Mastery')
    fireEvent.click(saveButton)

    await waitFor(() => {
      expect(getByText(/Invalid format/)).toBeInTheDocument()
    })
  })

  it('setting negative points sets error', async () => {
    const {getByText, getAllByLabelText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Set negative points
    const pointsInputs = getAllByLabelText(/Change points/)
    const firstInput = pointsInputs[0]
    firstInput.focus()
    firstInput.select()
    fireEvent.input(firstInput, {target: {value: '-1'}})

    // Submit form
    const saveButton = getByText('Save Learning Mastery')
    fireEvent.click(saveButton)

    await waitFor(() => {
      expect(getByText(/Negative points/)).toBeInTheDocument()
    })
  })

  it('sends POST on submit', async () => {
    const postSpy = jest
      .spyOn(axios, 'post')
      .mockImplementation(() => Promise.resolve({status: 200}))

    const {getByText} = render(<ProficiencyTable {...defaultProps} />)

    await wait(1)

    // Wait for billboard and close it
    await waitFor(() => {
      expect(getByText('Customize Learning Mastery Ratings')).toBeInTheDocument()
    })
    const getStartedButton = getByText('Get Started')
    fireEvent.click(getStartedButton)

    // Submit form
    const saveButton = getByText('Save Learning Mastery')
    fireEvent.click(saveButton)

    // Ensure that the mocked POST request was called
    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledTimes(1)
    })

    postSpy.mockRestore()
  })

  // Tests for error validation are covered by the UI interaction tests above
})

describe('custom proficiency', () => {
  it('renders two ratings that are deletable', async () => {
    const spy = jest.spyOn(axios, 'get').mockImplementation(() =>
      Promise.resolve({
        status: 200,
        data: {
          ratings: [
            {
              description: 'Great',
              points: 10,
              color: '0000ff',
              mastery: true,
            },
            {
              description: 'Poor',
              points: 0,
              color: 'ff0000',
              mastery: false,
            },
          ],
        },
      }),
    )

    const {getAllByRole} = render(<ProficiencyTable {...defaultProps} />)

    await waitFor(() => {
      const radios = getAllByRole('radio')
      expect(radios).toHaveLength(2)
    })

    const deleteButtons = getAllByRole('button', {name: /delete proficiency rating/i})
    expect(deleteButtons).toHaveLength(2)

    spy.mockRestore()
  })

  it('renders one rating that is not deletable', async () => {
    const spy = jest.spyOn(axios, 'get').mockImplementation(() =>
      Promise.resolve({
        status: 200,
        data: {
          ratings: [
            {
              description: 'Uno',
              points: 1,
              color: '0000ff',
              mastery: true,
            },
          ],
        },
      }),
    )

    const {getAllByRole, queryAllByLabelText} = render(<ProficiencyTable {...defaultProps} />)

    await waitFor(() => {
      const radios = getAllByRole('radio')
      expect(radios).toHaveLength(1)
    })

    const deleteButtons = queryAllByLabelText('Delete proficiency rating')
    expect(deleteButtons).toHaveLength(0)

    spy.mockRestore()
  })
})
