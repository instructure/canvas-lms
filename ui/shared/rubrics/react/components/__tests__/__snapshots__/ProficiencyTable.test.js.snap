// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default proficiency renders the ProficiencyRating component 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <Spinner
    as="div"
    margin="0 0 0 medium"
    renderTitle="Loading"
    size="large"
    variant="default"
  />
</div>
`;
